### DISCUSSION TITLE
Custom attributes for rpc metric

### DISCUSSION THREAD

**Discussion by chorin1** [open-telemetry/opentelemetry-java-instrumentation#8628] [Q&A]
When using the manual instrumentation for gRPC using `addServerAttributeExtractor` or `setCapturedServerRequestMetadata` adds the additional attributes only to the span and not to the `rpc.server.duration` metric.  
Is that intended?
Is it possible to add custom attributes for that metric?

**Comment by laurit** [comment]
Currently the list of attributes added to rpc metrics is hardcoded in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/rpc/MetricsView.java This was done to limit the cardinality of the metrics. We should devise a way to let users specify what attributes they want with with metrics views from the sdk https://github.com/open-telemetry/opentelemetry-java/tree/b14bed8d899d68d365ef590f5caf201ad95c0cb7/sdk-extensions/incubator

**Reply by chorin1** [reply]
Thanks for the quick answer. 
Without digging through the code, it was impossible to know what the `grpc-1.6` library provides and I found the [readme](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/grpc-1.6/library) extremely lacking. [instrumentation-api-semconv](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation-api-semconv) is even more cryptic in that regard. 
Regarding metrics:
As a user, I'd hope all the custom attributes added when building a `GrpcTelemetry` would also be included in the metrics (unless explicitly selected otherwise). Maybe the function names should reflect that or there should be additional functions for custom metric attributes in `GrpcTelemetryBuilder` -> `RpcServerMetrics` and `RpcClientMetrics` will be updated respectively to support that.

WDYT?
LMK if you want help with that.

**Reply by laurit** [reply]
What instrumentations provide is documented in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md see if that is what you were looking for.
There is a specification issue https://github.com/open-telemetry/opentelemetry-specification/issues/3061 for adding additional attributes to these metrics. Basically the ask is to provide an option for the agent to set the default attributes for these metrics that the user can override. How this is going to work for library users is still an open question as far as I understand. Perhaps your suggestion of letting user specify the attributes via `GrpcTelemetryBuilder` is the best option. @mateuszrzeszutek have you thought about how library users could set which metrics attributes thy want?

**Reply by trask** [reply]
I think(?) https://github.com/open-telemetry/opentelemetry-specification/issues/3061 would also solve the issue for library users. The grpc instrumentation would use the advice API to set the default attributes (corresponding to the semantic conventions), and then users could override this set of attributes by registering a metric view.

**Reply by mateuszrzeszutek** [reply]
> @mateuszrzeszutek have you thought about how library users could set which metrics attributes thy want?

I think whatever's extracted with an `AttributesExtractor` should be available to all signals, not just spans. The https://github.com/open-telemetry/opentelemetry-specification/issues/3061 is required to fix that though.
